### PR TITLE
Allow aeson-2.2

### PR DESCRIPTION
**Pull request by maksbotan**

### CODE REVIEW

**Comment by maksbotan** [comment]
@theophile-fl @ysangkok @jkarni and co, I need your advice on this. Servant.API.ContentTypes uses `Data.Aeson.Parser`:

https://github.com/haskell-servant/servant/blob/16cfd03e74557d8fd8ef6e2ebfb9c26bd645d756/servant/src/Servant/API/ContentTypes.hs#L374-L391

`Data.Aeson.Parser` module was removed in aeson-2.2, so there is no easy way to upgrade our version constraint.

However, I think that this function is not needed anymore. Haddock mentions that it parser any json value, contrary to aeson's behavior to parse only list and object. See haddock for 'json' in aeson-2.1: https://hackage.haskell.org/package/aeson-2.1.2.1/docs/Data-Aeson-Parser.html#v:json

> This function is an alias for [value](https://hackage.haskell.org/package/aeson-2.1.2.1/docs/Data-Aeson-Parser.html#v:value). In aeson 0.8 and earlier, it parsed only object or array types, in conformance with the now-obsolete RFC 4627.

Our lower bound on aeson is 1.4 for a long time, so I propose to remove our `eitherDecodeLenient` and use aeson's native `eitherDecode` instead. What do you think? I'm not sure if we have some tests to verify that the change is OK.

I will make the change in this PR a bit later.

**Comment by tchoutri** [comment]
I support your proposition, in any case we will want to make sure that a pre-release is uploaded to Hackage so that users (me included) can test the behaviour. :)

**Comment by maksbotan** [comment]
There was a test that servant's `eitherDecodeLenient` behaves exactly like `eitherDecode` from aeson:

https://github.com/haskell-servant/servant/blob/104a8ad2f387b68d6844101e01b585d106f4d0d3/servant/test/Servant/API/ContentTypesSpec.hs#L222-L229

I've pushed an update to check if everything works. I've also had to change the tutorial, as it also mentioned old `aeson` behavior.

File: servant/test/Servant/API/ContentTypesSpec.hs [hunk 1]
```diff
@@ -219,16 +219,6 @@ spec = describe "Servant.API.ContentTypes" $ do
                 handleCTypeH (Proxy :: Proxy '[JSONorText]) "image/jpeg"
                     "foobar" `shouldBe` (Nothing :: Maybe (Either String Int))
 
-    -- aeson >= 0.9 decodes top-level strings
-    describe "eitherDecodeLenient" $ do
-
-        it "parses top-level strings" $ do
```
Comment (ysangkok) [inline servant/test/Servant/API/ContentTypesSpec.hs:225]: It would be best if we could retain this test, ideally by having it use the `MimeUnrender JSON` instance that was modified. Just to make sure the docs are not mistaken.

**Comment by maksbotan** [comment]
I've returned the test

**Comment by maksbotan** [comment]
Do you want me to upload candidate to Hackage to test?

**Comment by tchoutri** [comment]
That would be lovely yes

**Comment by maksbotan** [comment]
It's here: https://hackage.haskell.org/package/servant-0.20.1/candidate

You can try it by adding https://hackage.haskell.org/package/servant-0.20.1/candidate/servant-0.20.1.tar.gz to your `cabal.project` or `stack.yaml`.

**Comment by tchoutri** [comment]
@maksbotan Wonderful. I think servant-server & servant-client-core are going to need a pre-release as well, I'm getting:

```

src/Servant/Server/Internal.hs:686:26: error:
    • Couldn't match type ‘a’ with ‘IO a0’
      Expected: SourceIO chunk -> a
        Actual: SourceIO chunk -> IO a0
      ‘a’ is a rigid type variable bound by
        the instance declaration
        at src/Servant/Server/Internal.hs:(673,5)-(675,68)
    • In the first argument of ‘return’, namely ‘fromSourceIO’
      In the expression: return fromSourceIO
      In an equation for ‘ctCheck’: ctCheck = return fromSourceIO
```

and

```
src/Servant/Client/Core/HasClient.hs:430:35: error:
    • Couldn't match type ‘a’ with ‘IO a1’
      Expected: Client m (Stream method status framing ct a)
        Actual: m (IO a1)
      ‘a’ is a rigid type variable bound by
        the instance declaration
        at src/Servant/Client/Core/HasClient.hs:(422,3)-(424,54)
    • In the expression:
        withStreamingRequest req'
          $ \ gres
              -> do let ...
                    return $ fromSourceIO $ framingUnrender' $ responseBody gres
      In an equation for ‘clientWithRoute’:
          clientWithRoute _pm Proxy req
            = withStreamingRequest req'
                $ \ gres
                    -> do let ...
                          ....
            where
                req'
                  = req
                      {requestAccept = fromList [...],
                       requestMethod = reflectMethod (Proxy :: Proxy method)}
      In the instance declaration for
        ‘HasClient m (Stream method status framing ct a)’
    • Relevant bindings include
        clientWithRoute :: Proxy m
                           -> Proxy (Stream method status framing ct a)
                           -> Request
                           -> Client m (Stream method status framing ct a)
          (bound at src/Servant/Client/Core/HasClient.hs:430:3)
    |
430 |   clientWithRoute _pm Proxy req = withStreamingRequest req' $ \gres -> do
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

```

**Comment by tchoutri** [comment]
I will also send a call for beta-testers on social media to make sure our users can take the time to test things on their own applications

**Comment by maksbotan** [comment]
Okay, so let's wait for a bit for more tests, and I'll merge & release.

**Comment by tchoutri** [comment]
@maksbotan servant-lucid would need a bound bump, I can do it for you if you give me permission on Hackage. :)

**Comment by ysangkok** [comment]
@maksbotan Has this been sufficiently tested already?

**Comment by Vekhir** [comment]
@maksbotan 
Is there anything blocking this from getting merged?

Publishing this patch in a new version would allow the Arch package to be updated without carrying an additional patch.

**Comment by larskuhtz** [comment]
I would also love to see this merged and released. 

(I think, as a web framework, servant should try to always support the latest version of aeson.)

File: .github/workflows/master.yml [hunk 2]
```diff
@@ -3,6 +3,8 @@ name: CI
 on:
   pull_request:
   push:
+    branches:
+      - master
```
Comment (larskuhtz) [inline .github/workflows/master.yml:7]: This seems unrelated to this PR.

File: .github/workflows/master.yml [hunk 3]
```diff
@@ -20,6 +22,7 @@ jobs:
           - "9.2.8"
           - "9.4.5"
           - "9.6.2"
+      fail-fast: false
```
Comment (larskuhtz) [inline .github/workflows/master.yml:25]: This change seems unrelated to this PR.

File: servant/src/Servant/API/ContentTypes.hs [hunk 4]
```diff
@@ -65,7 +65,6 @@ module Servant.API.ContentTypes
     , AllMime(..)
     , AllMimeRender(..)
     , AllMimeUnrender(..)
-    , eitherDecodeLenient
```
Comment (larskuhtz) [inline servant/src/Servant/API/ContentTypes.hs:68]: From the comments below it seems that with currently supported aeson versions `eitherDecodeLenient` and `eitherDecode` have the same behavior. 

What about making `eitherDecodeLenient` an alias of `eitherDecode` and just mark it as deprecated? That way it would not break the API.

File: servant/src/Servant/API/ContentTypes.hs [hunk 5]
```diff
@@ -371,28 +364,9 @@ instance NFData NoContent
 --------------------------------------------------------------------------
 -- * MimeUnrender Instances
```
Comment (larskuhtz) [inline servant/src/Servant/API/ContentTypes.hs:366]:
```suggestion
-- | Deprecated: since aeson version 0.9 `eitherDecode` has lenient behavior.
--
 eitherDecodeLenient :: FromJSON a => ByteString -> Either String a
 eitherDecodeLenient = eitherDecode
 {-# DEPRECATED eitherDecodeLenient "use eitherDecode instead" #-}
```

File: servant/src/Servant/API/ContentTypes.hs [hunk 6]
```diff
@@ -65,7 +65,6 @@ module Servant.API.ContentTypes
     , AllMime(..)
     , AllMimeRender(..)
     , AllMimeUnrender(..)
-    , eitherDecodeLenient
     , canHandleAcceptH
```
Comment (larskuhtz) [inline servant/src/Servant/API/ContentTypes.hs:68]: Keep `eitherDecodeLenient` as alias of `eitherDecde`, but mark it as deprecated (cf. below).

```suggestion
    , eitherDecodeLenient
    , canHandleAcceptH
```

File: servant/test/Servant/API/ContentTypesSpec.hs [hunk 7]
```diff
@@ -219,15 +219,13 @@ spec = describe "Servant.API.ContentTypes" $ do
                 handleCTypeH (Proxy :: Proxy '[JSONorText]) "image/jpeg"
                     "foobar" `shouldBe` (Nothing :: Maybe (Either String Int))
 
-    -- aeson >= 0.9 decodes top-level strings
     describe "eitherDecodeLenient" $ do
```
Comment (larskuhtz) [inline servant/test/Servant/API/ContentTypesSpec.hs:222]:
```suggestion
    describe "eitherDecode is lenient" $ do
```

**Review by larskuhtz — COMMENTED** [review]
I propose to keep `eitherDecodeLenient` as an deprecated alias of `eitherDecode`, since both have (probably) the same behavior since version 0.9 of aeson. (I haven't checked possible differences is the handling of of trailing characters, but I expect that to be a negligible issue.)

This would reduce the potential of the PR to break users and will hopefully help this change to be release soon.

**Comment by TeofilC** [comment]
Let me know if you'd appreciate help getting this over the line, but no pressure

**Comment by Vekhir** [comment]
The CI fails because it tries to build `warp-3.3.25` which is incompatible with `http2-4.2.0`. There is a newer version, `warp-3.3.29` which is compatible, but not chosen.

**Comment by maksbotan** [comment]
Thank you @Vekhir. I'll look into this.

**Comment by ysangkok** [comment]
> Thank you @Vekhir. I'll look into this.

You might want to remove or flip `constraints: warp < 3.3.26` in `cabal.project`.

**Comment by maksbotan** [comment]
Due to backwards-incompatible changes to master since this PR (see #1697), I will have to make Hackage release from this branch.

@ysangkok @tchoutri @larskuhtz please check out this release candidate: https://hackage.haskell.org/package/servant-0.20.1/candidate

Direct tar gz link: https://hackage.haskell.org/package/servant-0.20.1/candidate/servant-0.20.1.tar.gz

If everything's fine, I'll publish a release.

**Comment by tchoutri** [comment]
Alright, except for a usage of `eitherDecodeLenient` and odd-jobs lagging a bit behind for aeson-2.2 compat, I haven't encountered any problem.

**Comment by maksbotan** [comment]
What do you mean by "odd-jobs lagging"?

**Comment by tchoutri** [comment]
It still does not support aeson-2.2 https://github.com/saurabhnanda/odd-jobs/pull/101

**Comment by maksbotan** [comment]
Ah, didn't get that it's a name of a package :)

Great, I'll publish that release on Hackage.

**Comment by maksbotan** [comment]
Published!
